src/boot/bootctl-uki.c

```diff
@@ -30,12 +30,19 @@ static int pe_sections(FILE *uki, struct PeSectionHeader **ret, size_t *ret_n) {
         assert(ret);
         assert(ret_n);
 
-        items = fread(&dos, 1, sizeof(dos), uki);
-        if (items != sizeof(dos))
-                return log_error_errno(SYNTHETIC_ERRNO(EIO), "DOS header read error");
+        items = fread(&dos, 1, sizeof(dos.Magic), uki);
+        if (items != sizeof(dos.Magic))
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "File is smaller than DOS magic (got %"PRIu64" of %zu bytes)",
+                                       items, sizeof(dos.Magic));
         if (memcmp(dos.Magic, dos_file_magic, sizeof(dos_file_magic)) != 0)
                 goto no_sections;
 
+        rewind(uki);
+        items = fread(&dos, 1, sizeof(dos), uki);
```

Comment:
and then instead of rewinding here maybe do:

```c
items = fread((uint8_t*) &dos + items, 1, sizeof(DosFileHeader) - items, uki);
```

Reply:
The pointer arithmetic here is not exactly helpful to make the code more readable.
It avoids one syscall, but this certainly isn't a performance sensitive place ...

Reply:
Solved that slightly different now, do a single read but do two size checks, first magic only, then full header.